### PR TITLE
Add katello devel pulp integration pipeline

### PR DESCRIPTION
**Pull request by zjhuntin**

### CODE REVIEW

**Comment by zjhuntin** [comment]
Wasn't sure if I also needed to make the job in theforeman.org for it, apologies!

**Comment by evgeni** [comment]
> Wasn't sure if I also needed to make the job in theforeman.org for it, apologies!

only if you want to be able to trigger it from ci.tfm ;)

see https://github.com/theforeman/jenkins-jobs/blob/master/theforeman.org/yaml/jobs/pipeline/pulpcore-release-pipelines.yaml and https://github.com/theforeman/jenkins-jobs/blob/master/theforeman.org/pipelines/release/pipelines/pulpcore.groovy

You probably don't need the "version" magic, as your version is always "devel"

**Comment by zjhuntin** [comment]
@evgeni updated, I'm unsure what's causing the test error, any hints?

**Review by ekohl — COMMENTED** [review]
In https://github.com/theforeman/foreman-infra/pull/1535 I explicitly removed it because it was broken and nobody looked at it. Does this mean someone will look at the failures and resolve them if needed?

**Comment by evgeni** [comment]
the failure is:

```
2021-04-12T19:41:12.2891689Z INFO:jenkins_jobs.builder:Job name:  katello-pulpcore-integration-pipeline
2021-04-12T19:41:12.2893107Z DEBUG:jenkins_jobs.builder:Writing XML to '<_io.BufferedWriter name='<stdout>'>'
2021-04-12T19:41:12.2893865Z Traceback (most recent call last):
2021-04-12T19:41:12.2895915Z   File "/opt/hostedtoolcache/Python/3.9.4/x64/lib/python3.9/xml/etree/ElementTree.py", line 1039, in _escape_cdata
2021-04-12T19:41:12.2896794Z     if "&" in text:
2021-04-12T19:41:12.2897763Z TypeError: argument of type 'CustomLoaderCollection' is not iterable
2021-04-12T19:41:12.2898355Z 
2021-04-12T19:41:12.2898956Z During handling of the above exception, another exception occurred:
2021-04-12T19:41:12.2899479Z 
2021-04-12T19:41:12.2899899Z Traceback (most recent call last):
2021-04-12T19:41:12.2900969Z   File "/opt/hostedtoolcache/Python/3.9.4/x64/bin/jenkins-jobs", line 8, in <module>
2021-04-12T19:41:12.2901668Z     sys.exit(main())
2021-04-12T19:41:12.2902807Z   File "/opt/hostedtoolcache/Python/3.9.4/x64/lib/python3.9/site-packages/jenkins_jobs/cli/entry.py", line 174, in main
2021-04-12T19:41:12.2903648Z     jjb.execute()
2021-04-12T19:41:12.2904962Z   File "/opt/hostedtoolcache/Python/3.9.4/x64/lib/python3.9/site-packages/jenkins_jobs/cli/entry.py", line 154, in execute
2021-04-12T19:41:12.2906024Z     ext.obj.execute(self.options, self.jjb_config)
2021-04-12T19:41:12.2907412Z   File "/opt/hostedtoolcache/Python/3.9.4/x64/lib/python3.9/site-packages/jenkins_jobs/cli/subcommand/test.py", line 66, in execute
2021-04-12T19:41:12.2908379Z     builder.update_jobs(
2021-04-12T19:41:12.2909587Z   File "/opt/hostedtoolcache/Python/3.9.4/x64/lib/python3.9/site-packages/jenkins_jobs/builder.py", line 295, in update_jobs
2021-04-12T19:41:12.2910605Z     output.write(job.output())
2021-04-12T19:41:12.2911799Z   File "/opt/hostedtoolcache/Python/3.9.4/x64/lib/python3.9/site-packages/jenkins_jobs/xml_config.py", line 59, in output
2021-04-12T19:41:12.2913125Z     out = minidom.parseString(XML.tostring(self.xml, encoding="UTF-8"))
2021-04-12T19:41:12.2914267Z   File "/opt/hostedtoolcache/Python/3.9.4/x64/lib/python3.9/xml/etree/ElementTree.py", line 1109, in tostring
2021-04-12T19:41:12.2915436Z     ElementTree(element).write(stream, encoding,
2021-04-12T19:41:12.2916533Z   File "/opt/hostedtoolcache/Python/3.9.4/x64/lib/python3.9/xml/etree/ElementTree.py", line 748, in write
2021-04-12T19:41:12.2917521Z     serialize(write, self._root, qnames, namespaces,
2021-04-12T19:41:12.2918638Z   File "/opt/hostedtoolcache/Python/3.9.4/x64/lib/python3.9/xml/etree/ElementTree.py", line 913, in _serialize_xml
2021-04-12T19:41:12.2919588Z     _serialize_xml(write, e, qnames, None,
2021-04-12T19:41:12.2920659Z   File "/opt/hostedtoolcache/Python/3.9.4/x64/lib/python3.9/xml/etree/ElementTree.py", line 913, in _serialize_xml
2021-04-12T19:41:12.2921622Z     _serialize_xml(write, e, qnames, None,
2021-04-12T19:41:12.2922803Z   File "/opt/hostedtoolcache/Python/3.9.4/x64/lib/python3.9/xml/etree/ElementTree.py", line 911, in _serialize_xml
2021-04-12T19:41:12.2923750Z     write(_escape_cdata(text))
2021-04-12T19:41:12.2924741Z   File "/opt/hostedtoolcache/Python/3.9.4/x64/lib/python3.9/xml/etree/ElementTree.py", line 1047, in _escape_cdata
2021-04-12T19:41:12.2925699Z     _raise_serialization_error(text)
2021-04-12T19:41:12.2926834Z   File "/opt/hostedtoolcache/Python/3.9.4/x64/lib/python3.9/xml/etree/ElementTree.py", line 1029, in _raise_serialization_error
2021-04-12T19:41:12.2928106Z     raise TypeError(
2021-04-12T19:41:12.2929265Z TypeError: cannot serialize <jenkins_jobs.local_yaml.CustomLoaderCollection object at 0x7f1776a381c0> (type CustomLoaderCollection)
```

but why… is a great question

File: theforeman.org/pipelines/test/katello-pulpcore-integration.groovy [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+        ansiColor('xterm')
+    }
+
+    stages {
+        stage('Test Integration') {
+            agent any
+
+            steps {
+                script {
+                    runCicoPipelines('katello_devel', 'devel', pipelines)
```
Comment (evgeni) [inline theforeman.org/pipelines/test/katello-pulpcore-integration.groovy:17]: given you have no `version.groovy`, you'll need to define `pipelines` somehere, like in this file:

```
def pipelines = [
    'katello_devel': [
        'centos7',
        'centos8'
    ]
]
```

but that should not be the reason why it fails

**Comment by zjhuntin** [comment]
@evgeni I had copied this from a file using the job template, and didn't realize that if I listed all the dsl files with {empty} at the end, it'd recognize them as something it should iterate through, which caused the failure, ./generate_jobs.sh gave the error off nicely:

`TypeError: argument of type 'CustomLoaderCollection' is not iterable`

**Comment by evgeni** [comment]
Oh yeah, the `{empty}` hack is only needed when you do weird template stuff, which in this case you do not.

**Comment by evgeni** [comment]
So syntactically this is correct and will produce the jobs you want, but I have a few questions:

1. as Ewoud mentions, we had this, and it was red for ages, who is going to ensure it's not this time?
2. this PR says it's about pulpcore, but the respective pipeline in forklift is a generic katello_devel one, without any specific pulpcore bits?

**Comment by zjhuntin** [comment]
This one is somewhat different as they've added it to test the pulpcore integration testing alongside it which evidently is something that the Katello team working on pulpcore integration was wanting run once a day, but let me also get James in on this conversation as I think he'll be able to point out the particulars and answer other questions on it.

**Comment by jjeffers** [comment]
> So syntactically this is correct and will produce the jobs you want, but I have a few questions:
> 
> 1. as Ewoud mentions, we had this, and it was red for ages, who is going to ensure it's not this time?
> 2. this PR says it's about pulpcore, but the respective pipeline in forklift is a generic katello_devel one, without any specific pulpcore bits?

1. If no one else, I will, but since this was a Katello "Team Content" request then the team has an interest in this pipeline.
2. It's probably dumb luck on my part - when I provisioned this locally, I did have pulpcore enabled in the virtual machine, etc. If I missed a step, I apologize.

**Comment by evgeni** [comment]
>     2\. It's probably dumb luck on my part - when I provisioned this locally, I did have pulpcore enabled in the virtual machine, etc. If I missed a step, I apologize.

Well, no, it's deploying a "katello-devel" box, which is using pulpcore as that's the artifact storage katello uses these days. So it's totally intentional. I was just wondering why this name, as pulpcore is the default anyways.

**Comment by jjeffers** [comment]
>  I was just wondering why this name, as pulpcore is the default anyways.

The name might be a lie of omission. The playbook invokes a subset of Katello/pulpcore integration tests in "live" mode. Typically the tests use the pre-recorded VCR cassettes to mock interactions with Pulpcore. The live mode essentially forces the tests to use the actual installed client responses and not the recorded cassette responses.

**Review by evgeni — APPROVED** [review]
I *personally* find the name "pulpcore-integration" confusing, as it's not *limited* to pulpcore (I'd just call it `katello-devel-pipeline`), but at the same time, I won't be the one watching this, so I don't mind.

If @katello is happy with that naming, and will monitor that job, enjoy!

**Comment by zjhuntin** [comment]
I think katello-devel-pipeline makes more sense in general for it, @jjeffers you OK if I change it to that?

**Comment by jjeffers** [comment]
> I _personally_ find the name "pulpcore-integration" confusing, as it's not _limited_ to pulpcore (I'd just call it `katello-devel-pipeline`), but at the same time, I won't be the one watching this, so I don't mind.
> 
> If @Katello is happy with that naming, and will monitor that job, enjoy!

The test invocation _only_ calls the pulpcore tests in live mode. The vast majority of katello tests are not involved.

**Comment by evgeni** [comment]
> The test invocation _only_ calls the pulpcore tests in live mode. The vast majority of katello tests are not involved.

Then the pipeline name in forklift seems wrong, but honestly, merge this, clean it later.

**Comment by evgeni** [comment]
So, according to jenkins, since we merged this, https://ci.theforeman.org/job/katello-pulpcore-integration-pipeline/ was not green once.

Is that being looked into?

**Comment by jjeffers** [comment]
> So, according to jenkins, since we merged this, https://ci.theforeman.org/job/katello-pulpcore-integration-pipeline/ was not green once.
> 
> Is that being looked into?

That's me, I'll take a look at it.

**Comment by evgeni** [comment]
@jjeffers, according to jenkins, https://ci.theforeman.org/job/katello-pulpcore-integration-pipeline/ still wasn't ever green, and looking at the past failures, it didn't even get past the installer phase.

so my question is: is this pipeline still useful to y'all or can we re-purpose the resources?

**Comment by evgeni** [comment]
@jjeffers montly ping on this ;)

**Comment by ekohl** [comment]
Revert PR is here: https://github.com/theforeman/jenkins-jobs/pull/191.

**Pull request merged**